src/Utilities/DoubleDicts.jl

```diff
@@ -3,116 +3,100 @@ module DoubleDicts
 import MathOptInterface
 
 const MOI = MathOptInterface
-const MOIU = MOI.Utilities
-const CI = MOI.ConstraintIndex
-const AD = AbstractDict
 
-const DataTypePair = Tuple{DataType,DataType}
+abstract type AbstractDoubleDict{V} <: AbstractDict{MOI.ConstraintIndex,V} end
 
-abstract type AbstractDoubleDict{K,V,IK,DI<:AD{IK},DO<:AD{K,DI}} <: AD{CI,V} end
+abstract type AbstractDoubleDictInner{F,S,V} <:
+              AbstractDict{MOI.ConstraintIndex{F,S},V} end
+
+"""
+    typed_value(dict::AbstractDoubleDictInner{F,S,V}, value) where {F,S,V}
+
+Convert the `value` stored inside `dict` to the equivalent on the outer
+`DoubleDict`. This is useful when the value type `V` of the inner dict is
+different to the outer dict. (See, e.g., [`IndexDoubleDict`](@ref).)
+"""
+function typed_value(::AbstractDoubleDictInner{F,S,V}, value::V) where {F,S,V}
+    return value
+end
 
 """
     DoubleDict{V}
 
-Optimized dictionary to map `ConstraintIndex` (`CI`) to values of type `V`.
-Works as a `AbstractDict{CI, V}` with minimal differences.
-Note that `CI` is not a concrete type, opposed to `CI{MOI.SingleVariable, MOI.Integers}`,
-which is a concrete type.
+An optimized dictionary to map `MOI.ConstraintIndex` to values of type `V`.
+
+Works as a `AbstractDict{MOI.ConstraintIndex,V}` with minimal differences.
+
+If `V` is also a `MOI.ConstraintIndex`, use [`IndexDoubleDict`](@ref).
 
-When optimal performance or type stability is required it is possible to obtain a
-fully type stable dictionary with values of type `V` and keys of type
-`CI{MOI.SingleVariable, MOI.Integers}` from the dictionary `dict`, for instance:
+Note that `MOI.ConstraintIndex` is not a concrete type, opposed to
+`MOI.ConstraintIndex{MOI.SingleVariable, MOI.Integers}`, which is a concrete
+type.
 
-    ``inner = dict[MOI.SingleVariable, MOI.Integers]``
+When looping through multiple keys of the same Function-in-Set type, use
+```julia
+inner = dict[F, S]
+```
+to return a type-stable [`DoubleDictInner`](@ref).
 """
-struct DoubleDict{V,DI,DO} <: AbstractDoubleDict{DataTypePair,V,Int64,DI,DO}
-    dict::DO
+struct DoubleDict{V} <: AbstractDoubleDict{V}
+    dict::Dict{Tuple{DataType,DataType},Dict{Int64,V}}
```

Comment:
This way we wont be able to use clever dicts inside double dicts.
Which might be ok for now. But if variables need that for performance, chances are that constraints will need to.

Reply:
Same thing on the keys side... if we want to generalize that later we might be in trouble.
I completely understand the point of cleaning up and adding thing back later. Wouldnt this way break if we bring back functionality?
If it wont break, it is less of an issue.
Other option is to make the fully parametric dd as a underscore type. While the Most used IndexDD is no undersocored.

Reply:
I'll consider making the dict a type parameter if there is a clear real-world benchmark showing that this is a bottleneck. 

If we need some uber-DoubleDict-CleverDict, we can always add that in future as a new subtype of `AbstractDoubleDict`. 

For now, DoubleDicts solves a clear use-case: a ConstraintIndex->ConstaintIndex map in IndexMap. Let's keep that simplicity.